crates/wasmparser/src/validator/core.rs

```diff
@@ -494,14 +495,71 @@ pub(crate) struct Module {
 impl Module {
     pub fn add_type(
         &mut self,
-        ty: crate::Type,
+        ty: SubType,
         features: &WasmFeatures,
         types: &mut TypeAlloc,
         offset: usize,
         check_limit: bool,
     ) -> Result<()> {
-        let ty = match ty {
-            crate::Type::Func(t) => {
+        if check_limit {
+            check_max(self.types.len(), 1, MAX_WASM_TYPES, "types", offset)?;
+        }
+        let ty = self.check_subtype(ty, features, types, offset)?;
+
+        let id = types.push_ty(ty);
+        self.types.push(id);
+        Ok(())
+    }
+
+    fn check_subtype(
+        &mut self,
+        ty: SubType,
+        features: &WasmFeatures,
+        types: &mut TypeAlloc,
+        offset: usize,
+    ) -> Result<Type> {
+        if let Some(type_index) = ty.supertype_idx {
```

Comment:
I think this may need a feature check for `gc`?

Reply:
Added.